apps/passport-server/src/apis/lemonade/lemonadeAPI.ts

```diff
@@ -1,78 +1,553 @@
+  oauthClientSecret: string;
+  oauthServerUrl: string;
 }
 
 /**
- * A lemonade ticket as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
+ * Matches the structure of Lemonade's GraphQL API.
  */
-export interface LemonadeTicket {
-  id: string;
-  name: string;
-  email: string;
-  eventId: string;
-  tierId: string;
-  checkedIn: boolean;
-  checkerEmail?: string;
+export interface ILemonadeAPI {
+  getHostingEvents(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials
+  ): Promise<LemonadeEvent[]>;
+
+  getEventTicketTypes(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicketTypes>;
+
+  getTickets(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicket[]>;
+
+  checkinUser(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string,
+    lemonadeUserId: string
+  ): Promise<LemonadeCheckin>;
 }
 
 /**
- * A lemonade ticket tier as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
+ * To enable different auth token generation strategies in tests, provide an
+ * interface that allows for swappable token generators.
  */
-export interface LemonadeTicketTier {
-  name: string;
-  id: string;
+interface AuthTokenSource {
+  getToken(credentials: LemonadeOAuthCredentials): Promise<string>;
 }
 
 /**
- * A lemonade event as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
+ * In the OAuth token manager, we want to cache certain objects between calls,
+ * as they contain reusable tokens or reusable OAuth data that can be used to
+ * regenerate expired tokens.
  */
-export interface LemonadeEvent {
-  id: string;
-  name: string;
-  tickets: LemonadeTicket[];
-  tiers: LemonadeTicketTier[];
-  permissionedUserIds: string[];
+interface CachedTokenIssuance {
+  issuer: Issuer;
+  client: Client;
+  tokenSet: TokenSet;
 }
 
 /**
- * TODO: implement to match the actual API
+ * Manages tokens fetched from an OAuth backend.
  */
-export interface ILemonadeAPI {
-  loadEvents(apiKey: string): Promise<LemonadeEvent[]>;
-  checkinTicket(
-    apiKey: string,
-    eventId: string,
-    ticketId: string
-  ): Promise<void>;
-  // TODO: fill in the other methods. This is what is planned so far:
-  // - API Key scoped to an ‘account’, which has read/write permissions to
-  //   the appropriate events. E.g. can read/write events that are owned/co-owned by the account.
-  // - GET product types for a given event
-  // - GET tickets for a given event, which will include
-  //     - Attendee name
-  //     - Attendee email
-  //     - Product type (7-day, early bird, GA, etc)
-  //     - Checked-in status
-  // - POST Check-in (and potentially check-out)
+class OAuthTokenManager implements AuthTokenSource {
```

Comment:
The API object is a singleton, and it ensures that only one `OAuthTokenManager` is instantiated for any set of credentials (it's roughly analogous to how the Pretix API object ensures that there's only one rate-limit queue per organizer).

Reply:
Oh, no, I seem to have changed that. Hmm.

Reply:
It is certainly possible that different pipelines could have different credentials, but since they all share the same LemonadeAPI object, they would not be independently generating tokens.